src/platform/datapath_win.c

```diff
@@ -0,0 +1,636 @@
+        NewSocket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatSocketCreateTcpListener(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_opt_ const QUIC_ADDR* LocalAddress,
+    _In_opt_ void* RecvCallbackContext,
+    _Out_ CXPLAT_SOCKET** NewSocket
+    )
+{
+    return SocketCreateTcpListener(
+        Datapath,
+        LocalAddress,
+        RecvCallbackContext,
+        NewSocket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatSocketDelete(
+    _In_ CXPLAT_SOCKET* Socket
+    )
+{
+    if (Socket->RawSocketAvailable) {
+        RawSocketDelete(SocketToRaw(Socket));
+    }
+    SocketDelete(Socket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatSocketUpdateQeo(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_reads_(OffloadCount)
+        const CXPLAT_QEO_CONNECTION* Offloads,
+    _In_ uint32_t OffloadCount
+    )
+{
+    if (Socket->UseTcp || (Socket->RawSocketAvailable &&
+        !IS_LOOPBACK(Offloads[0].Address))) {
+        return RawSocketUpdateQeo(SocketToRaw(Socket), Offloads, OffloadCount);
+    }
+    return QUIC_STATUS_NOT_SUPPORTED;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+UINT16
+CxPlatSocketGetLocalMtu(
+    _In_ CXPLAT_SOCKET* Socket
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    if (Socket->UseTcp || (Socket->RawSocketAvailable &&
+        !IS_LOOPBACK(Socket->RemoteAddress))) {
+        return RawSocketGetLocalMtu(SocketToRaw(Socket));
+    }
+    return Socket->Mtu;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSocketGetLocalAddress(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Out_ QUIC_ADDR* Address
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    *Address = Socket->LocalAddress;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSocketGetRemoteAddress(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Out_ QUIC_ADDR* Address
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    *Address = Socket->RemoteAddress;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatRecvDataReturn(
+    _In_opt_ CXPLAT_RECV_DATA* RecvDataChain
+    )
+{
+    if (RecvDataChain == NULL) {
+        return;
+    }
+
+    if (RecvDataChain->DatapathType == CXPLAT_DATAPATH_TYPE_USER) {
+        RecvDataReturn(RecvDataChain);
+    } else if (RecvDataChain->DatapathType == CXPLAT_DATAPATH_TYPE_XDP) {
+        RawRecvDataReturn(RecvDataChain);
+    } else {
+        CXPLAT_DBG_ASSERT(FALSE);
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Success_(return != NULL)
+CXPLAT_SEND_DATA*
+CxPlatSendDataAlloc(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Inout_ CXPLAT_SEND_CONFIG* Config
+    )
+{
+    CXPLAT_SEND_DATA* SendData = NULL;
+    // TODO: fallback?
+    if (Socket->UseTcp || Config->Route->DatapathType == CXPLAT_DATAPATH_TYPE_XDP ||
+        (Config->Route->DatapathType == CXPLAT_DATAPATH_TYPE_UNKNOWN &&
+        Socket->RawSocketAvailable && !IS_LOOPBACK(Config->Route->RemoteAddress))) {
+        SendData = RawSendDataAlloc(SocketToRaw(Socket), Config);
+        if (SendData) {
```

Comment:
Instead of setting the Type fields here, update the two SendDataAlloc functions to set the type.